client/web/src/repo/blob/codemirror/token-selection/decorations.ts

```diff
@@ -0,0 +1,103 @@
+                    if (state.field(isModifierKeyHeld) && hasDefinition && focus.occurrence === hover?.occurrence) {
+                        classes.push('cm-token-selection-definition-ready')
+                    }
+
+                    decorations.push({
+                        decoration: Decoration.mark({
+                            class: classes.join(' '),
+                            attributes: {
+                                // Selected (focused) occurrence is the only focusable element in the editor.
+                                // This helps to maintain the focus position when editor is blurred and then focused again.
+                                tabindex: '0',
+                            },
+                        }),
+                        range: focus.occurrence.range,
+                    })
+
+                    // Decorate selected (focused) occurrence document highlights.
+                    const highlights = state.field(documentHighlightsField)
+                    const focusedOccurrenceHighlight = findByOccurrence(highlights, focus.occurrence)
+                    if (focusedOccurrenceHighlight) {
+                        for (const highlight of sortRangeValuesByStart(highlights)) {
+                            if (highlight === focusedOccurrenceHighlight) {
+                                // Focused occurrence is already highlighted.
+                                continue
+                            }
+
+                            decorations.push({
+                                decoration: Decoration.mark({
+                                    class: 'sourcegraph-document-highlight',
+                                }),
+                                range: highlight.range,
+                            })
+                        }
+                    }
+                }
+
+                if (pin) {
+                    decorations.push({
+                        decoration: Decoration.mark({ class: 'selection-highlight' }),
+                        range: pin.occurrence.range,
+                    })
+                }
+
+                if (hover) {
+                    decorations.push({
+                        decoration: Decoration.mark({ class: 'selection-highlight' }),
+                        range: hover.occurrence.range,
+                    })
+                }
+
+                const ranges = decorations.reduce((acc, { decoration, range }) => {
+                    const from = positionToOffset(state.doc, range.start)
+                    const to = positionToOffset(state.doc, range.end)
+
+                    if (from !== null && to !== null) {
+                        acc.push(decoration.range(from, to))
+                    }
+
+                    return acc
+                }, [] as Range<Decoration>[])
+
+                return Decoration.set(ranges.sort((a, b) => a.from - b.from))
```

Comment:
Minor nit but we can move the sorting function into the module scope so we only need to allocate it once

```suggestion
                return Decoration.set(ranges.sort(sortByFrom)
```